File: src/lessons/bat/string1/Last2Entity.java
package lessons.bat.string1;

import static plm.core.ValueSerializer.*;

import plm.universe.bat.BatEntity;

public class Last2Entity extends BatEntity {

  public void run()
  {
    int count = getTestCount();
    for (int i = 0; i < count; i++) {
      Object[] param = (Object[])deserialize(getTest(i));
      setTestResult(i, serialize(last2((String)param[0])));
    }
  }

  /* BEGIN TEMPLATE */
  int last2(String str)
  {
    /* BEGIN SOLUTION */
    // Screen out too-short string case.
    if (str.length() < 2)
      return 0;

    String end = str.substring(str.length() - 2);
    // Note: substring() with 1 value goes through the end of the string
    int count = 0;

    // Check each substring length 2 starting at i
    for (int i = 0; i < str.length() - 2; i++) {
      String sub = str.substring(i, i + 2);
      if (sub.equals(end)) { // Use .equals() with strings
        count++;
      }
    }

    return count;
    /* END SOLUTION */
  }
  /* END TEMPLATE */
}

File: src/lessons/bat/string1/Last2Entity.scala
package lessons.bat.string1

import plm.core.ValueSerializer
import plm.universe.bat.BatEntity

class Last2Entity extends BatEntity {

    override def run() {
      import ValueSerializer._

      val count = getTestCount()
      for (i <- 0 to count -1) {
        val param = deserialize(getTest(i)).asInstanceOf[Array[Object]]
        setTestResult(i, serialize(last2(param(0).asInstanceOf[String])) )
    }
	}

	/* BEGIN TEMPLATE */
	def last2(str:String):Int = {
  	/* BEGIN SOLUTION */
    val l = str.length
    if (l < 2)
      return 0
    val end = str.substring(l-2,l)
    var count = 0
    for (i <- 0 to str.length-3)
      if (str.substring(i,i+2) == end)
        count += 1
    return count
  	/* END SOLUTION */
	}
	/* END TEMPLATE */
}
